Core indent rule reports false positives in doc blocks

**What version of TypeScript are you using?**
`typescript@2.6.2`

**What version of `typescript-eslint-parser` are you using?**
`typescript-eslint-parser@11.0.0`

**What code were you trying to parse?**

```ts
/**
 * This is a function.
 * @param {String} text
 * @returns {String}
 */
function foo(text: string): string {}

```

The minimal repro case is:
```ts
/**
 * This is a function.
 * @param {String} text
 * @returns {String}
 */
function foo() {}
```

Running the autofixer generates the following code (which then passes):
```ts
 /**
 * This is a function.
{String} text
{String}
 */
function foo(text: string): string {}
```

**What did you expect to happen?**
No errors! This only occurs with the typescript-eslint-parser. It does not happen with the default parser.

**What happened?**
```sh
/Users/cataldo/Code/test/eslint-test/index.js
  3:1  error  Expected indentation of 0 spaces but found 3  indent
  4:1  error  Expected indentation of 0 spaces but found 3  indent
```

I'll try to look into the cause if I get a chance, but not sure if and when that'll be. Took the bug label off because I'm not sure what the established practice is in this repo - happy to update though :)

--- a/lib/node-utils.js
+++ b/lib/node-utils.js
@@ -245,6 +245,32 @@ function isComma(token) {
 }
 
 /**
+ * Returns true if the given TSToken is a comment
+ * @param  {TSToken}  token the TypeScript token
+ * @returns {boolean}       is commment
+ */
+function isComment(token) {
+    return (token.kind === SyntaxKind.SingleLineCommentTrivia || token.kind === SyntaxKind.MultiLineCommentTrivia) || (token.kind >= SyntaxKind.JSDocTypeExpression && token.kind <= SyntaxKind.JSDocTypeLiteral);
+}
+
+/**
+ * Returns true if the given TSToken is inside a comment
+ * Non-comment type tokens are generated for types in JSDoc Blocks
+ * @param  {TSToken}  token the TypeScript token
+ * @returns {boolean}       is inside a commment
+ */
+function isInsideComment(token) {
+    if (token.kind === SyntaxKind.SourceFile) {
+        return false;
+    }
+    if (isComment(token)) {
+        return true;
+    }
+    return isInsideComment(token.parent);
+}
+
+
+/**
  * Returns the binary expression type of the given TSToken
  * @param  {TSToken} operator the operator token
  * @returns {string}          the binary expression type
@@ -693,7 +719,7 @@ function convertTokens(ast) {
      * @returns {undefined}
      */
     function walk(node) {
-        if (isToken(node) && node.kind !== SyntaxKind.EndOfFileToken) {
+        if (isToken(node) && !isInsideComment(node) && node.kind !== SyntaxKind.EndOfFileToken) {
             const converted = convertToken(node, ast);
 
             if (converted) {

--- a/lib/node-utils.js
+++ b/lib/node-utils.js
@@ -185,7 +185,9 @@ module.exports = {
     convertTokens,
     getNodeContainer,
     isWithinTypeAnnotation,
-    isTypeKeyword
+    isTypeKeyword,
+    isComment,
+    isJSDocComment
 };
 /* eslint-enable no-use-before-define */
 
@@ -245,30 +247,22 @@ function isComma(token) {
 }
 
 /**
- * Returns true if the given TSToken is a comment
- * @param  {TSToken}  token the TypeScript token
- * @returns {boolean}       is commment
+ * Returns true if the given TSNode is a comment
+ * @param {TSNode} node the TypeScript node
+ * @returns {boolean} is commment
  */
-function isComment(token) {
-    return (token.kind === SyntaxKind.SingleLineCommentTrivia || token.kind === SyntaxKind.MultiLineCommentTrivia) || (token.kind >= SyntaxKind.JSDocTypeExpression && token.kind <= SyntaxKind.JSDocTypeLiteral);
+function isComment(node) {
+    return node.kind === SyntaxKind.SingleLineCommentTrivia || node.kind === SyntaxKind.MultiLineCommentTrivia;
 }
 
 /**
- * Returns true if the given TSToken is inside a comment
- * Non-comment type tokens are generated for types in JSDoc Blocks
- * @param  {TSToken}  token the TypeScript token
- * @returns {boolean}       is inside a commment
+ * Returns true if the given TSNode is a JSDoc comment
+ * @param {TSNode} node the TypeScript node
+ * @returns {boolean} is JSDoc comment
  */
-function isInsideComment(token) {
-    if (token.kind === SyntaxKind.SourceFile) {
-        return false;
-    }
-    if (isComment(token)) {
-        return true;
-    }
-    return isInsideComment(token.parent);
+function isJSDocComment(node) {
+    return node.kind === SyntaxKind.JSDocComment;
 }
-
 
 /**
  * Returns the binary expression type of the given TSToken
@@ -282,7 +276,6 @@ function getBinaryExpressionType(operator) {
         return "LogicalExpression";
     }
     return "BinaryExpression";
-
 }
 
 /**
@@ -719,7 +712,13 @@ function convertTokens(ast) {
      * @returns {undefined}
      */
     function walk(node) {
-        if (isToken(node) && !isInsideComment(node) && node.kind !== SyntaxKind.EndOfFileToken) {
+        // TypeScript generates tokens for types in JSDoc blocks. Comment tokens
+        // and their children should not be walked or added to the resulting tokens list.
+        if (isComment(node) || isJSDocComment(node)) {
+            return;
+        }
+
+        if (isToken(node) && node.kind !== SyntaxKind.EndOfFileToken) {
             const converted = convertToken(node, ast);
 
             if (converted) {

--- a/tests/fixtures/comments/jsdoc-comment.src.js
+++ b/tests/fixtures/comments/jsdoc-comment.src.js
@@ -1,17 +1,4 @@
 /**
- * @a
- */
-foo;
-
-/**
- * @a
- */
-/**
- * a
- */
-foo;
-
-/**
  * This is a function.
  * @param {String} bar some string
  * @returns {String} returns bar

--- a/tests/integration/external-fixtures/jsdoc-indent.js
+++ b/tests/integration/external-fixtures/jsdoc-indent.js
@@ -10,3 +10,12 @@ foo;
  * a
  */
 foo;
+
+/**
+ * This is a function.
+ * @param {String} bar
+ * @returns {String} returns bar
+ */
+function foo(bar) {
+    return bar;
+}

--- a/tests/integration/typescript.spec.js
+++ b/tests/integration/typescript.spec.js
@@ -80,7 +80,7 @@ describe("TypeScript", () => {
         );
     });
 
-    it("should not produce any lint errors on valid JSDoc indentation (#344)", () => {
+    it("should not produce any lint errors on valid JSDoc indentation (#344 & #422)", () => {
         verifyAndAssertMessages(
             loadExternalFixture("jsdoc-indent"),
             {

